### PR TITLE
Upgrade TSLint to ESLint, rename package tslint -> lint

### PR DESCRIPTION
**Pull request by ryanio**
TSLint is now [deprecated](https://www.npmjs.com/package/tslint) in favor of ESLint.

This PR is a __breaking change__, and does the following:

* Renames package from `ethereumjs-config-tslint` to `ethereumjs-config-lint`
   * As per https://github.com/ethereumjs/ethereumjs-config/issues/18
   * This will also need to be updated in npm
* Uses recommended rules from [`typestrict`](https://github.com/krzkaczor/TypeStrict), `eslint`, [`typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint) and `prettier`

### How repos using this package should update

Once published in npm, update in `package.json`:

```
-    "@ethereumjs/config-tslint": "^1.1.1",
+    "@ethereumjs/config-lint": "^2.0.0",
```

Add `.eslintrc.js` with:

```js
module.exports = {
  parserOptions: {
    project: './tsconfig.json',
  }
};
```

Use bin cmd:

`ethereumjs-config-lint`

---

This PR also aims to close https://github.com/ethereumjs/ethereumjs-config/issues/19 and https://github.com/ethereumjs/ethereumjs-vm/issues/640

### CODE REVIEW

**Comment by holgerd77** [comment]
Hi Ryan,
I think naming packages after the tools was a mistake in the first place. This should rather be named after functionality to become tool-agnostic, so names should be something like:

- ethereumjs-config-nyc -> ethereumjs-config-coverage
- ethereumjs-config-prettier -> ethereumjs-config-formatting
- ethereumjs-config-tsc	-> ethereumjs-config-typescript
- ethereumjs-config-tslint -> ethereumjs-config-linting

If you agree: I think it would be worth to do the switch, maybe always along some breaking releases. So do we want to start here on this occasion with the `linting`  package?

**Comment by ryanio** [comment]
Yeah sounds great, I think it’s a good idea.

I think without the “ing” is fine so “-format” and “-lint”, if you agree. 

Should we review and merge this then I’ll do a separate PR for the rest?

**Comment by ryanio** [comment]
Also, for this PR, should I port over the existing rules on top of the recommended base set? 

I’m not sure the history of choosing the existing rule set.

**Comment by holgerd77** [comment]
We can very well leave the "ing". 🙂 Review and merge this first also sounds fine.

Not sure on the rule set - this is the first time I get in touch with this TSLint -> ESLint deprecation - how much do rules differ between the two?

**Comment by ryanio** [comment]
> Not sure on the rule set - this is the first time I get in touch with this TSLint -> ESLint deprecation - how much do rules differ between the two?

I migrated the existing rules from `tslint` to `eslint` using [this guide](https://github.com/typescript-eslint/typescript-eslint/blob/be3e23f940608ea32bc804617ca01a95e456b491/packages/eslint-plugin/ROADMAP.md). Based on `git blame` it looks like most of the tslint file was set up by @krzkaczor and you later added `no-console` and `no-commented-code`.

This is the current `tslint`:

```json
{
  "extends": "typestrict",
  "rules": {
    "no-console": {
      "severity": "warning"
    },
    "no-debugger": true,
    "mocha-avoid-only": true,
    "prefer-const": true,
    "no-var-keyword": true,
    "interface-name": [true, "never-prefix"],
    "no-commented-code": {
      "severity": "warning"
    },
    "no-use-before-declare": false
  }
}
```

The biggest thing we're missing in eslint is `typestrict`, however it looks like `typescript-eslint` recommended rules has done a lot of work to ensure type checking [here](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin#recommended-configs).

I was able to convert all of the existing rules except `mocha-avoid-only` and `no-commented-code` (found someone's custom rule for `no-commented-code` but seems like there may be some open issues with it [here](https://github.com/bahmutov/eslint-rules/issues/17)).

This is the resulting `eslint`:

```js
module.exports = {
  parser: "@typescript-eslint/parser",
  plugins: ["@typescript-eslint"],
  env: {
    es6: true,
    node: true
  },
  ignorePatterns: ["node_modules/**/*", "**/node_modules/**/*", "dist/**/*"],
  extends: [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking",
    "plugin:prettier/recommended"
  ],
  rules: {
    "no-console": "warn",
    "no-debugger": "error",
    "prefer-const": "error",
    "no-var": "error",
    "@typescript-eslint/no-use-before-define": "error",
    "@typescript-eslint/interface-name-prefix": [
      "error",
      {
        prefixWithI: "never"
      }
    ]
  },
  parserOptions: {
    sourceType: "module"
  }
};
```

**Comment by holgerd77** [comment]
Ok, great!

`typestrict`: I think we might be fine with the better integrated type checking then

`no-commented-code`: no strong feeling there, would be nice and somewhat quality-improving to have this in, but also not the end of the world to leave out

Generally: this was our very first try to have some more explicit unification and agreement on the linting rules. Don't have anything specific in mind and this generally worked well. We can nevertheless very much do some between conclusion here, improve on stuff or remove things which didn't work so well. Feel very free to add your own recommendations to the discussion. We were a limited group of people discussing this and there is very much room for change and additional voices, eventually @evertonfraga and/or @cgewecke also want to have a word here.

I wouldn't want to merge here before EthCC I think, we can let this sink in a bit further. This is also a good topic to discuss and exchange about at EthCC face-to-face. Feel free to do some continued work here though if necessary.

File: packages/ethereumjs-config-nyc/package.json [hunk 1]
```diff
@@ -15,5 +15,9 @@
   },
   "publishConfig": {
     "access": "public"
+  },
+  "peerDependencies": {
+    "nyc": "^15.0.0",
```
Comment (evertonfraga) [inline packages/ethereumjs-config-nyc/package.json]: Please use version `< 15` (`^13.2.0` was fine), to produce reports with absolute paths, which is important for our current monorepo setting.

Comment (ryanio) [reply packages/ethereumjs-config-nyc/package.json]: I will update, thanks.

**Comment by krzkaczor** [comment]
FYI: TypeStrict just migrated to ESLint! https://github.com/krzkaczor/TypeStrict/issues/19

**Comment by ryanio** [comment]
Amazing, thanks @krzkaczor! I will update this PR to include it 😎

**Comment by holgerd77** [comment]
@ryanio just give an explicit comment once this is open again for review, might be overlooked otherwise

**Comment by ryanio** [comment]
@holgerd77 @krzkaczor ok, just pushed some updates and this is ready for another review!

I tried to do a test run in `ethereumjs-vm/packages/vm` by first running `npm link` in `ethereumjs-config` and then in the vm dir: `npm link @ethereumjs/config-lint`. It linted successfully so I think it worked, but I'll have to throw an error in on purpose to make sure it's working as intended :)

File: packages/ethereumjs-config-lint/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+module.exports = {
+  parser: "@typescript-eslint/parser",
+  plugins: ["@typescript-eslint"],
+  env: {
+    es6: true,
+    node: true
+  },
+  ignorePatterns: ["node_modules/**/*", "**/node_modules/**/*", "dist/**/*"],
```
Comment (holgerd77) [inline packages/ethereumjs-config-lint/.eslintrc.js:8]: Why is here this double asterisk path for `node_modules` but not for `dist`?

Comment (holgerd77) [inline packages/ethereumjs-config-lint/.eslintrc.js:8]: Why is here this double asterisk path for `node_modules` but not for `dist`?

Comment (evertonfraga) [reply packages/ethereumjs-config-lint/.eslintrc.js:8]: The `dist` directories that are under a module's dependencies would be excluded by the first rule already.

**Comment by holgerd77** [comment]
What is the status here?

File: packages/ethereumjs-config-lint/.eslintrc.js (hunk 2 above)
Comment (ryanio) [reply packages/ethereumjs-config-lint/.eslintrc.js:8]: @evertonfraga thanks, i will update :)

**Comment by holgerd77** [comment]
@ryanio Thanks for the update! 😄 Is this now ready for review?

**Comment by holgerd77** [comment]
(overall status is a bit confusing since the current change request is still active and displayed)

**Comment by ryanio** [comment]
Hm thanks yes @evertonfraga can you update your review if the updated code looks good?

I am going to try to use `verdaccio` or `npm link` to insert the package into ethereumjs-vm and see if the linting is working well, and if I can throw a lint error on purpose. After that I'd be comfortable moving ahead with merge + npm release on the new name :)

**Comment by evertonfraga** [comment]
Sorry, I totally missed this notification. Will resume with testing tomorrow!

**Comment by ryanio** [comment]
ok I made some updates and I got `eslint` successfully running on `ethereumjs-vm/packages/vm`.

This is what I needed to do:

1. cd `ethereumjs-config`
    1. checkout branch `upgradeToEslint`
    1. run `npm install`
    1. run `npm link`
2. cd `ethereumjs-vm/packages/vm`
    1. remove `@ethereumjs/config-tslint` from package.json and add dep `"eslint": "^7.0.0"`
    1. run `npm link @ethereumjs/config-link`
    1. create a `.eslintrc.js` file with the following:

```js
module.exports = {
  parserOptions: {
    project: './tsconfig.json',
  }
};
```

and run `npm run lint`

I got 97 lint problems however they seem to be just `import/require` and `var/const/let` failures that can be `fix`ed along with the upgrade in `vm`.

The only last thing I would like to figure out if there is a way to not need `typescript`,`eslint`, and `prettier` as deps in this package. there is a  `--resolve-plugins-relative-to` flag for eslint but then all the `@typescript-eslint` deps would need to be installed into every project.

**Comment by evertonfraga** [comment]
Right, thanks for the work here @ryanio! It took me a little while get the "whole picture" of this package and its current state.

I am testing with VM, made some changes in `package.json` to pretend it was a fresh package. The UX can still improve, but nothing specific to this PR.

These are the current imposed requirements to make the `ethereumjs-config-lint` script work:
1. have an `.eslintrc.js` file
1. add script `"format": "ethereumjs-config-format"` to package.json
1. add script `"tsc": "ethereumjs-config-typescript"`to package.json

I made some basic error handling for (1). But I have a feeling that we must do something about the other bullet points. Suggestions:
- Add `prettier` and `tsc` to the package, so we can just run the commands directly, instead of relying in the consumer `package.json`. 

- Make `lint` less powerful — yet more granular. It would only execute `ESLint`, and the consumer libraries would be able to choose the mix of tools it needs to use.  Example:
```js
"scripts": {
  "codecheck": "ethereumjs-config-lint && ethereumjs-config-format"
}
```

**Comment by evertonfraga** [comment]
I just checked: `Lint` is the only package of this project that runs commands from sister packages, creating this coupling issue.

My gut feeling tells me to drop the `prettier/format` and `tsc` from `lint` and move on. This PR is already a breaking change, anyway.

**Comment by ryanio** [comment]
> My gut feeling tells me to drop the `prettier/format` and `tsc` from `lint` and move on. This PR is already a breaking change, anyway :)

that sounds good to me!

**Comment by evertonfraga** [comment]
Ok, I'll work on that now

**Comment by ryanio** [comment]
@evertonfraga looks great, thanks for the rebase!

**Comment by holgerd77** [comment]
Not really able to catch up with all the developments here, but from first sight things seem to be going in a good direction, so just another 👍 to go ahead here once you are confident on the changes and merge.

**Comment by evertonfraga** [comment]
Thanks! I'd like to make a test run on all tasks today. This series of PRs have touched every internal package.

File: .gitignore [hunk 3]
```diff
@@ -1,4 +1,2 @@
 node_modules
-package-lock.json
-packages/*/**/node_modules
-packages/*/**/package-lock.json
+package-lock.json
```
Comment (evertonfraga) [inline .gitignore]: 👍

**Comment by evertonfraga** [comment]
By the way, I made [this little script](https://gist.github.com/evertonfraga/a6653ad10ac27b62a0805f856168658a#file-migrate-ethereumjs-config-sh), to help migrate the package changes. There's another one to help link the packages using `file:` protocol.

As anything I have to do in the monorepo is x6, this kind of thing is super handy to me.

**Comment by evertonfraga** [comment]
@ryanio I followed your [instructions from here](https://github.com/ethereumjs/ethereumjs-config/pull/26#issuecomment-630520246), and I got the same result as you:

> I got 97 lint problems however they seem to be just import/require and var/const/let failures that can be fixed along with the upgrade in vm.

But the 97 errors described indicate that ESLint is not configured properly, as it presents parser errors, as it does not extend the one defined in `lint` package.

>   1:1  error  Parsing error: The keyword 'const' is reserved

**Comment by evertonfraga** [comment]
Additionally, I found some eslint plugins installation mismatch. 

Can you provide updated information on how to setup this plugin? 

```npm WARN eslint-config-typestrict@1.0.0 requires a peer of @typescript-eslint/eslint-plugin@^2.25.0 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-plugin-sonarjs@0.5.0 requires a peer of eslint@^3.0.0 || ^4.0.0 || ^5.0.0 || ^6.0.0 but none is installed. You must install peer dependencies yourself.
npm WARN eslint-plugin-import@2.2.0 requires a peer of eslint@2.x - 3.x but none is installed. You must install peer dependencies yourself.
npm WARN eslint-plugin-react@6.10.3 requires a peer of eslint@^2.0.0 || ^3.0.0 but none is installed. You must install peer dependencies yourself.```

**Comment by ryanio** [comment]
@evertonfraga thanks for your patience with this PR 😄

This is currently working for me with latest updates to branch, if you want to give it a try:

In `ethereumjs-config/packages/lint` run `npm link`

In `ethereumjs-vm/packages/account`:

Run `npm link @ethereumjs/config-lint` 

Create `.eslintrc.js`:
```js
module.exports = {
  extends: '@ethereumjs/eslint-config-helper',
  parserOptions: {
    project: './tsconfig.eslint.json',
  },
}
```

Create `tsconfig.eslint.json`:
```js
{
  "extends": "./tsconfig.json",
  "include": ["src/**/*.ts", "test/**/*.ts", ".eslintrc.js", "prettier.config.js", "typedoc.js"]
}
```

(See https://github.com/typescript-eslint/typescript-eslint/issues/1723#issuecomment-626766041 for why `tsconfig.eslint.json` is needed.)

In account, run something like `node_modules/eslint/bin/eslint.js --fix --config ./.eslintrc.js . --resolve-plugins-relative-to ~/dev/ethereumjs-config/packages/lint`

(`--resolve-plugins-relative-to` is needed otherwise the consumer lib will need to install all the plugins used. I have tested adding this to lint.sh with `--resolve-plugins-relative-to node_modules/@ethereumjs/config-lint`, I hope it works)

My results:

```
/Users/ryanghods/dev/ethereumjs-vm/packages/account/test/index.spec.ts
  152:43  error  'err' is defined but never used                                             no-unused-vars
  152:43  error  'err' is defined but never used. Allowed unused args must match /^_/u       @typescript-eslint/no-unused-vars
  152:48  error  'codeHash' is defined but never used                                        no-unused-vars
  152:48  error  'codeHash' is defined but never used. Allowed unused args must match /^_/u  @typescript-eslint/no-unused-vars
  200:43  error  'err' is defined but never used                                             no-unused-vars
  200:43  error  'err' is defined but never used. Allowed unused args must match /^_/u       @typescript-eslint/no-unused-vars
```

**Comment by evertonfraga** [comment]
Sure, I can rename it now. 

I am testing it and doing a test integration to the VM.

**Comment by holgerd77** [comment]
@evertonfraga just so that there won't stay any confusion: do you left this for @ryanio to merge or is there anything explicitly blocking this or other reasoning to leave this open for now?

**Comment by evertonfraga** [comment]
I am making integration tests with the monorepo and will merge this one soon

**Pull request merged**